lib/Conversion/Affine/AffineFullUnroll.cpp

```diff
@@ -0,0 +1,29 @@
+#include "lib/Conversion/Affine/AffineFullUnroll.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/LoopUtils.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/include/mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace tutorial {
+
+using mlir::affine::AffineForOp;
+using mlir::affine::loopUnrollFull;
+
+struct UnrollAffineLoopPattern : public OpRewritePattern<AffineForOp> {
+  explicit UnrollAffineLoopPattern(MLIRContext *context)
+      : OpRewritePattern(context, 1) {}
+
+  LogicalResult matchAndRewrite(AffineForOp op, PatternRewriter &rewriter) const override {
+    return loopUnrollFull(op);
+  }
+};
+
+void AffineFullUnrollPass::runOnOperation() {
+  RewritePatternSet patterns(&getContext());
+  patterns.add<UnrollAffineLoopPattern>(&getContext());
+  (void)applyPatternsAndFoldGreedily(getOperation(), std::move(patterns));
```

Comment:
Since we were talking about threading, I'm not actually sure that the rewrite/conversion drivers parallelize! It seems like the passes are, so for a given pass you can't modify outside the root operation (in this case Func).

I'm peeking around at the implementations for pattern applications and rewrites: https://github.com/llvm/llvm-project/blob/43dce27c06e49ba98543d447734f41f454256216/mlir/lib/Rewrite/PatternApplicator.cpp#L126

Reply:
TIL, Thanks!